src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,665 @@
+          sigma_tilde_s.array() = pre_factor * sigma_tilde_s.array()
+                                + post_factor
+                                  * sigma_tilde_grad.array().square();
+
+          // Take ADAgrad or rmsprop step
+          musigmatilde.set_mu(
+            musigmatilde.mu().array() +
+            eta_stepsize_ * mu_grad.array()
+            / (tau + mu_s.array().sqrt()));
+          musigmatilde.set_sigma_tilde(
+            musigmatilde.sigma_tilde().array()  +
+            eta_stepsize_ * sigma_tilde_grad.array()
+            / (tau + sigma_tilde_s.array().sqrt()));
+
+          // Check for convergence every "refresh_"th iteration
+          if (iter_counter % refresh_ == 0) {
+            elbo_prev = elbo;
+            elbo = calc_ELBO(musigmatilde);
+            delta_elbo = rel_decrease(elbo, elbo_prev);
+            cb.push_back(delta_elbo);
+            delta_elbo_ave = std::accumulate(cb.begin(), cb.end(), 0.0)
+                             / (double)(cb.size());
+            delta_elbo_med = circ_buff_median(cb);
+            *print_stream_
+                      << "  "
+                      << std::setw(4) << iter_counter
+                      << "  "
+                      << std::right << std::setw(9) << std::setprecision(1)
+                      << elbo
+                      << "  "
+                      << std::setw(16) << std::fixed << std::setprecision(3)
+                      << delta_elbo_ave
+                      << "  "
+                      << std::setw(15) << std::fixed << std::setprecision(3)
+                      << delta_elbo_med;
+
+            if (diag_stream_) {
+              end = clock();
+              delta_t = static_cast<double>(end - start) / CLOCKS_PER_SEC;
+
+              print_vector.clear();
+              print_vector.push_back(delta_t);
+              print_vector.push_back(elbo);
+              services::io::write_iteration_csv(*diag_stream_,
+                                                iter_counter, print_vector);
+            }
+
+            if (delta_elbo_ave < tol_rel_obj) {
+              *print_stream_ << "   MEAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (delta_elbo_med < tol_rel_obj) {
+              *print_stream_ << "   MEDIAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (iter_counter > 100) {
+              if (delta_elbo_med > 0.5 || delta_elbo_ave > 0.5) {
+                *print_stream_ << "   MAY BE DIVERGING... INSPECT ELBO";
+              }
+            }
+
+            *print_stream_ << std::endl;
+          }
+
+          // Check for max iterations
+          if (iter_counter == max_iterations) {
+            *print_stream_ << "MAX ITERATIONS REACHED" << std::endl;
+            do_more_iterations = false;
+          }
+
+          ++iter_counter;
+        }
+      }
+
+      void run_fullrank(double tol_rel_obj, int max_iterations) {
```

Comment:
@akucukelbir Do we need this function? Doesn't seem to do much aside from calling two other functions.